Add data preparation utility for hierarchical time series

rshyamsundar opened the pull request:
*Issue #, if available:*

*Description of changes:*

This PR implements the missing functionality for handling hierarchical/grouped time series data and training hierarchical models correctly without data-preprocessing issues (see #1967 and #2067).  It also provides a tutorial on loading hierarchical time series and training and evaluation of the hierarchical model [`DeepVARHierarchical`](https://github.com/awslabs/gluon-ts/blob/dev/src/gluonts/mx/model/deepvar_hierarchical/_estimator.py#L95).

**Details:**
* The user has to provide only the bottom-level time series (pandas dataframe) and the aggregation matrix (numpy ndarray), which are the minimal requirements in constructing hierarchical or even grouped time series.  The aggregated time series are then internally constructed and all the time series in the hierarchy are available as a pandas dataframe.

* The hierarchical time series thus constructed can be easily converted to a `Dataset` on which mini-batch training can be done, thanks to the newly implemented [`PandasDataset`](https://github.com/awslabs/gluon-ts/blob/dev/src/gluonts/dataset/pandas.py#L27).

* The tutorial explains the procedure for creating hierarchical time series, training the model and doing evaluation via backtesting by creating the train-test split explicitly (i.e., not using `make_evaluation_predictions` or `backtest_metrics`).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.


**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

rshyamsundar committed:
Add data preparation utility for hierarchical time series

rshyamsundar committed:
Add tutorial notebook

rshyamsundar committed:
Fix typo

rshyamsundar committed:
Merge branch 'dev' into hier_data_prep

rshyamsundar committed:
Fix flake

rshyamsundar committed:
Fix style checks

rshyamsundar committed:
Fix pytests

lostella commented:
@rshyamsundar you also need to add the tutorial to the table of contents in the `index.md` in the same directory as the `.md.template` file

rshyamsundar committed:
Add tutorial name in index.md

rshyamsundar committed:
Merge branch 'dev' into hier_data_prep

lostella commented on docs/tutorials/forecasting/hierarchical_model_tutorial.md.template:
```diff
@@ -0,0 +1,177 @@
+# Make sure the dataframe has `PeriodIndex`.
+ts_at_bottom_level = ts_at_bottom_level.to_period()
+
+# Load the aggregation matrix `S`.
+S_csv = (
+    "https://gist.githubusercontent.com/rshyamsundar/17084fd1f28021867bcf6f2d69d9b73a/raw/"
+    "32780ca43f57a78f2d521a75e73b136b17f34a02/example_agg_mat.csv"
+)
+S = pd.read_csv(S_csv).values
```
Would it make sense to break this cell into multiple ones, to stop and explain what is being dealt with? For example, maybe showing what `S` contains and what it means would make this easier to follow or adapt to one's use case.

rshyamsundar replied:
`S` is the standard summation matrix that anyone using the hierarchical models is familiar with. Also it is the user input and not an artifact of `gluonts` code. That's why I didn't want to dwell on it too much. We can provide this [reference link (Eq. 10.3)](https://otexts.com/fpp2/hts.html#fig:HierTree) to make it clear the expected format of the summation matrix.

lostella commented on docs/tutorials/forecasting/hierarchical_model_tutorial.md.template:
````diff
@@ -0,0 +1,177 @@
+Here we convert it into `gluonts.dataset.pandas.PandasDataset`.
+
+```python
+from gluonts.dataset.hierarchical import to_pandas_dataset
+
+dataset = to_pandas_dataset(hts=hts)
````
This could be a method of `HierarchicalTimeSeries`:

```python
hts.to_pandas_dataset()
```

I'm wondering whether `pandas` should be in the name here though: in the end, what matters to the user is that this is converting a `HierarchicalTimeSeries` object into a `Dataset`, so maybe `to_dataset` is a better name.

rshyamsundar replied:
ya, actually it is better it if it a method of `HierarchicalTimeSeries`.

I tried being explicit about to which dataset I am converting to (`PandasDataset`). `to_dataset` is also fine and we don't imagine converting this to to other types of datasets anyway right?

rshyamsundar replied:
Now I remember why I moved `to_pandas_dataset` out of the class. We are not converting just the `hts` to `Dataset` here. We are taking external features and the target time series (happened to be `hts`) to create a dataset. So I thought `to_dataset(hts, features)` is better than `hts.to_dataset(features)`. Your call!

lostella replied:
Right, but features are optional, so I’d say making it a method is fine

rshyamsundar replied:
Moved.

lostella commented on docs/tutorials/forecasting/hierarchical_model_tutorial.md.template:
````diff
@@ -0,0 +1,177 @@
+    S=S,
+)
+
+# You can access all the time series of the hierarchy using `ts_at_all_levels` property.
+hts.ts_at_all_levels.head()
+```
````
Also here: instead of using comments, you could break the cell and explain this in a text section

rshyamsundar replied:
Done

lostella commented on docs/tutorials/forecasting/hierarchical_model_tutorial.md.template:
```diff
@@ -0,0 +1,177 @@
+)
+S = pd.read_csv(S_csv).values
+
+# Create `HierarchicalTimeSeries` object.
```
This comment seems a bit redundant :-)

rshyamsundar replied:
Removed.

lostella commented on docs/tutorials/forecasting/hierarchical_model_tutorial.md.template:
```diff
@@ -0,0 +1,177 @@
+    "https://gist.githubusercontent.com/rshyamsundar/39e57075743537c4100a716a7b7ec047/"
+    "raw/f02f5aeadad73e3f3e9cf54478606d3507826492/example_bottom_ts.csv"
+)
+ts_at_bottom_level = pd.read_csv(
+    ts_at_bottom_level_csv,
+    index_col=0,
+    parse_dates=True,
+)
+
+# Make sure the dataframe has `PeriodIndex`.
+ts_at_bottom_level = ts_at_bottom_level.to_period()
```
```suggestion
ts_at_bottom_level = pd.read_csv(
    ts_at_bottom_level_csv,
    index_col=0,
    parse_dates=True,
).to_period()
```

The fact that a `PeriodIndex` is needed could be stressed in the text

rshyamsundar replied:
I had your suggested version initially and thought people might miss `to_period` call if it is in a single statement and raise unnecessary issues around this point. So made that explicit. I will change and stress it in the text.

rshyamsundar replied:
Done

lostella commented on docs/tutorials/forecasting/hierarchical_model_tutorial.md.template:
````diff
@@ -0,0 +1,177 @@
+hts.ts_at_all_levels.head()
+```
+
+## Use case 1 - Simple Model Training and Forecasting
````
Not sure these two are orthogonal use cases, maybe we can just title this 

```suggestion
## Model Training and Forecasting
```

rshyamsundar replied:
Done

lostella commented on docs/tutorials/forecasting/hierarchical_model_tutorial.md.template:
````diff
@@ -0,0 +1,177 @@
+```
+
+
+## Use case 2 - Model Evaluation via Backtesting
````
Similarly to above:

```suggestion
## Model Evaluation via Backtesting
```

rshyamsundar replied:
Done

lostella commented on src/gluonts/dataset/hierarchical.py:
```diff
@@ -0,0 +1,217 @@
+    S
+        Summation or aggregation matrix whose ordering should be consistent
+        with the orderig of the columns of `ts_at_all_levels`.
+        In particular, the bottom `k x k` sub-matrix should be identity matrix,
+        where `k` is the number of leaves of the hierarchy.
```
If this is strictly required, then maybe it's fine to omit the bottom `k x k` submatrix and leave it implicit?

That would mean that `S` encodes the non-leaf nodes in the hierarchy only. I understand this may go against common mathematical notation, but it seems better on the coding side of things.

rshyamsundar replied:
Didn't want to introduce yet another notation and users typically have this matrix available for their own processing tasks as well: multiplying the forecasts at bottom level by this matrix gives them "all" forecasts in one go. There is a good reason for the introduction of `S` in the literature and from the user-perspective, it makes sense to use this standard matrix.

lostella replied:
I still think it's strange to require some input if part of it is expected to have specific values. But we can go on with this and maybe revisit it later.

rshyamsundar committed:
Address comments on the notebook

melopeo commented:
I think this PR is in great shape. One thing that could make it even better is to include an example (maybe like a 3rd section after `Model Training and Forecasting`, and `Model Evaluation via Backtesting`), where `feat_dynamic_real` are included. I suggest this as it is one element from `gluonts.dataset.hierarchical.to_pandas_dataset` that remains a tiny bit unclear how to use, and one can expect questions about this in the future.

rshyamsundar committed:
make to_dataset a class method

rshyamsundar committed:
Fix style checks

rshyamsundar committed:
Fix flake

rshyamsundar committed:
Merge branch 'dev' into hier_data_prep

rshyamsundar commented:
> I think this PR is in great shape. One thing that could make it even better is to include an example (maybe like a 3rd section after `Model Training and Forecasting`, and `Model Evaluation via Backtesting`), where `feat_dynamic_real` are included. I suggest this as it is one element from `gluonts.dataset.hierarchical.to_pandas_dataset` that remains a tiny bit unclear how to use, and one can expect questions about this in the future.

Hi @melopeo Here is the [PR](https://github.com/awslabs/gluon-ts/pull/2253) that addresses your comments. Thanks!
